File: lib/milestones.ts
const FIXED_MILESTONES = [
  { days: 90, label: "90 days" },
  { days: 180, label: "6 months" },
  { days: 270, label: "9 months" },
  { days: 365, label: "1 year" },
  { days: 545, label: "18 months" },
];

export type Milestone = {
  days: number;
  label: string;
  previousDays: number;
};

/** 90 days, 6/9/18 months, 1 year, then annually — matches docs/mvp-scope.md → Milestones. */
function* milestoneSequence(): Generator<{ days: number; label: string }> {
  yield* FIXED_MILESTONES;

  let years = 2;
  let target = 730;
  while (true) {
    yield { days: target, label: `${years} years` };
    target += 365;
    years += 1;
  }
}

export function getNextMilestone(daysSober: number): Milestone {
  let previousDays = 0;

  for (const milestone of milestoneSequence()) {
    if (daysSober < milestone.days) {
      return { ...milestone, previousDays };
    }
    previousDays = milestone.days;
  }

  throw new Error("unreachable");
}

/** Every milestone already reached, in order — each one earns a badge. */
export function getEarnedMilestones(daysSober: number): Milestone[] {
  const earned: Milestone[] = [];
  let previousDays = 0;

  for (const milestone of milestoneSequence()) {
    if (milestone.days > daysSober) break;
    earned.push({ ...milestone, previousDays });
    previousDays = milestone.days;
  }

  return earned;
}

export function milestoneProgress(daysSober: number, milestone: Milestone): number {
  const span = milestone.days - milestone.previousDays;
  if (span <= 0) return 100;
  return Math.min(100, Math.max(0, ((daysSober - milestone.previousDays) / span) * 100));
}

File: ios/RhythmCore/Sources/RhythmCore/Milestones.swift
import Foundation

/// A milestone target with the span it covers, for progress-bar math.
/// `previousDays` is the prior milestone's `days` (0 for the first milestone).
public struct Milestone: Equatable, Sendable {
    public let days: Int
    public let label: String
    public let previousDays: Int

    public init(days: Int, label: String, previousDays: Int) {
        self.days = days
        self.label = label
        self.previousDays = previousDays
    }
}

private let fixedMilestones: [(days: Int, label: String)] = [
    (90, "90 days"),
    (180, "6 months"),
    (270, "9 months"),
    (365, "1 year"),
    (545, "18 months"),
]

/// 90 days, 6/9/18 months, 1 year, then annually — matches docs/mvp-scope.md → Milestones.
/// `next()` on the returned iterator is non-mutating on the wrapper itself: the running
/// state lives inside the captured closure, matching the web's generator function.
private func milestoneSequence() -> AnyIterator<(days: Int, label: String)> {
    var index = 0
    var years = 2
    var target = 730
    return AnyIterator {
        if index < fixedMilestones.count {
            defer { index += 1 }
            return fixedMilestones[index]
        }
        let milestone = (days: target, label: "\(years) years")
        target += 365
        years += 1
        return milestone
    }
}

public func getNextMilestone(daysSober: Int) -> Milestone {
    let iterator = milestoneSequence()
    var previousDays = 0
    while let milestone = iterator.next() {
        if daysSober < milestone.days {
            return Milestone(days: milestone.days, label: milestone.label, previousDays: previousDays)
        }
        previousDays = milestone.days
    }
    fatalError("unreachable")
}

/// Every milestone already reached, in order — each one earns a badge.
public func getEarnedMilestones(daysSober: Int) -> [Milestone] {
    let iterator = milestoneSequence()
    var earned: [Milestone] = []
    var previousDays = 0
    while let milestone = iterator.next() {
        if milestone.days > daysSober { break }
        earned.append(Milestone(days: milestone.days, label: milestone.label, previousDays: previousDays))
        previousDays = milestone.days
    }
    return earned
}

/// Progress toward `milestone`, as a percentage in `0...100`.
public func milestoneProgress(daysSober: Int, milestone: Milestone) -> Double {
    let span = milestone.days - milestone.previousDays
    if span <= 0 { return 100 }
    let raw = (Double(daysSober - milestone.previousDays) / Double(span)) * 100
    return min(100, max(0, raw))
}
